lib/Sema/ConstraintGraph.cpp

```diff
@@ -1558,6 +1558,119 @@ void ConstraintGraph::dump(llvm::raw_ostream &out) {
   print(CS.getTypeVariables(), out);
 }
 
+void ConstraintGraph::print(llvm::raw_ostream &out, unsigned indent) {
+  if (Changes.empty() && (ActiveScope->getStartIdx() == Changes.size()))
+    return;
+
+  // Collect Changes for printing.
+  std::map<TypeVariableType *, TypeBase *> tvWithboundTypes;
+  std::vector<TypeVariableType *> addedTypeVars;
+  std::vector<TypeVariableType *> equivTypeVars;
+  std::vector<Constraint *> addedConstraints;
+  std::vector<Constraint *> removedConstraints;
+  for (unsigned int i = ActiveScope->getStartIdx(); i < Changes.size(); i++) {
+    auto change = Changes[i];
+    switch (change.Kind) {
+    case ChangeKind::BoundTypeVariable:
+      tvWithboundTypes.insert(std::pair<TypeVariableType *, TypeBase *>(
+          change.Binding.TypeVar, change.Binding.FixedType));
+      break;
+    case ChangeKind::AddedTypeVariable:
+      addedTypeVars.push_back(change.TypeVar);
+      break;
+    case ChangeKind::ExtendedEquivalenceClass:
+      equivTypeVars.push_back(change.EquivClass.TypeVar);
+      break;
+    case ChangeKind::AddedConstraint:
+      addedConstraints.push_back(change.TheConstraint);
+      break;
+    case ChangeKind::RemovedConstraint:
+      removedConstraints.push_back(change.TheConstraint);
+      break;
+    }
+  }
+
+  // If there are any constraints that were both added and removed in this set
+  // of Changes, remove them from both.
+  for (std::vector<Constraint *>::const_iterator itr = addedConstraints.begin();
+       itr != addedConstraints.end();) {
+    auto repeatedConstraintItr =
```

Comment:
There is an easy way to do this - store add/removed in sets instead of vectors, produce a union of them and remove all the constraints from that union.